File: codetree/2025_2.py
from math import ceil

lamps = []
lamps_num = []
last_m = 0
n, m = 0, 0


def add_lamp():
    global last_m
    global m
    max_len, max_i = 0, 0
    for i in range(m - 1):
        len = lamps[i + 1] - lamps[i]
        if len > max_len:
            max_len = len
            max_i = i
    lamps.insert(max_i + 1, ceil((lamps[max_i] + lamps[max_i + 1]) / 2))
    lamps_num.insert(max_i + 1, last_m + 1)
    last_m += 1
    m += 1


def delete_lamp(target):
    global m
    for i in range(len(lamps)):
        if lamps_num[i] == target:
            lamps.pop(i)
            lamps_num.pop(i)
            break
    m -= 1


def get_r():
    global m
    max_len = max(n - lamps[m - 1], lamps[0] - 1) * 2
    for i in range(m - 1):
        leng = lamps[i + 1] - lamps[i]
        if max_len < leng:
            max_len = leng
    if max_len < 0:
        return 0
    return max_len


q = int(input())
for _ in range(q):
    command = list(map(int, input().split()))
    if command[0] == 100:
        n, m = command[1], command[2]
        lamps = command[3:]
        lamps_num = [i for i in range(1, m + 1)]
        last_m = m
    elif command[0] == 200:
        add_lamp()
    elif command[0] == 300:
        delete_lamp(command[1])
    elif command[0] == 400:
        print(get_r())

File: codetree/2025_2.cpp
#include <iostream>
#include <set>
#include <map>
using namespace std;

set<int> lamps;
multiset<int> gaps;
map<int, int> lamp_num_to_pos;

int last_m = 0;
int n = 0, m = 0;

void add_gap(int a, int b) {
    gaps.insert(b - a);
}

void erase_gap(int a, int b) {
    multiset<int>::iterator it = gaps.find(b - a);
    if (it != gaps.end()) {
        gaps.erase(it);
    }
}

void add_lamp() {
    if (lamps.size() < 2) return;

    int max_len = -1;
    set<int>::iterator max_left = lamps.begin();

    set<int>::iterator it = lamps.begin();
    set<int>::iterator next_it = it;
    ++next_it;

    while (next_it != lamps.end()) {
        int len = *next_it - *it;
        if (len > max_len) {
            max_len = len;
            max_left = it;
        }
        ++it;
        ++next_it;
    }

    int left = *max_left;
    set<int>::iterator right_it = max_left;
    ++right_it;
    int right = *right_it;

    int new_pos = (left + right + 1) / 2; // ceil((left + right) / 2.0)

    erase_gap(left, right);

    lamps.insert(new_pos);
    ++last_m;
    lamp_num_to_pos[last_m] = new_pos;

    add_gap(left, new_pos);
    add_gap(new_pos, right);

    ++m;
}

void delete_lamp(int target) {
    if (lamp_num_to_pos.find(target) == lamp_num_to_pos.end()) return;

    int target_pos = lamp_num_to_pos[target];
    set<int>::iterator it = lamps.find(target_pos);

    if (it == lamps.end()) return;

    set<int>::iterator prev_it = it;
    set<int>::iterator next_it = it;
    bool has_prev = false;
    bool has_next = false;

    if (it != lamps.begin()) {
        --prev_it;
        has_prev = true;
    }

    ++next_it;
    if (next_it != lamps.end()) {
        has_next = true;
    }

    if (has_prev) {
        erase_gap(*prev_it, target_pos);
    }
    if (has_next) {
        erase_gap(target_pos, *next_it);
    }
    if (has_prev && has_next) {
        add_gap(*prev_it, *next_it);
    }

    lamps.erase(it);
    lamp_num_to_pos.erase(target);

    --m;
}

int get_r() {
    if (m == 0) return 0;

    int left_edge = (*lamps.begin() - 1) * 2;
    int right_edge = (n - *lamps.rbegin()) * 2;

    int max_len = 0;
    if (!gaps.empty()) {
        max_len = *gaps.rbegin();
    }

    int answer = left_edge;
    if (answer < right_edge) answer = right_edge;
    if (answer < max_len) answer = max_len;

    return answer;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    cin >> q;

    while (q--) {
        int command;
        cin >> command;

        if (command == 100) {
            cin >> n >> m;

            lamps.clear();
            gaps.clear();
            lamp_num_to_pos.clear();

            for (int i = 1; i <= m; i++) {
                int x;
                cin >> x;
                lamps.insert(x);
                lamp_num_to_pos[i] = x;
            }

            last_m = m;

            if (lamps.size() >= 2) {
                set<int>::iterator it = lamps.begin();
                set<int>::iterator next_it = it;
                ++next_it;

                while (next_it != lamps.end()) {
                    add_gap(*it, *next_it);
                    ++it;
                    ++next_it;
                }
            }
        }
        else if (command == 200) {
            add_lamp();
        }
        else if (command == 300) {
            int target;
            cin >> target;
            delete_lamp(target);
        }
        else if (command == 400) {
            cout << get_r() << '\n';
        }
    }

    return 0;
}